OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs

```diff
@@ -745,5 +748,78 @@ void MakeMouseFocusSettingsLive()
 			else
 				Game.Renderer.ReleaseWindowMouseFocus();
 		}
+
+		void InitHotkeyRemapDialog()
+		{
+			hotkeyDialogRoot.Get<LabelWidget>("HOTKEY_LABEL").GetText = () => selectedHotkeyDefinition.Description + ":";
+
+			duplicateHotkeyNotice = hotkeyDialogRoot.Get<LabelWidget>("DUPLICATE_NOTICE");
+			duplicateHotkeyNotice.TextColor = ChromeMetrics.Get<Color>("NoticeErrorColor");
+			duplicateHotkeyNotice.GetText = () =>
+			{
+				return (duplicateHotkeyDefinition != null) ? duplicateHotkeyNotice.Text.F(duplicateHotkeyDefinition.Description) : duplicateHotkeyNotice.Text;
+			};
+
+			defaultHotkeyNotice = hotkeyDialogRoot.Get<LabelWidget>("DEFAULT_NOTICE");
+			defaultHotkeyNotice.TextColor = ChromeMetrics.Get<Color>("NoticeInfoColor");
+
+			originalHotkeyNotice = hotkeyDialogRoot.Get<LabelWidget>("ORIGINAL_NOTICE");
+			originalHotkeyNotice.TextColor = ChromeMetrics.Get<Color>("NoticeInfoColor");
+			originalHotkeyNotice.GetText = () => originalHotkeyNotice.Text.F(selectedHotkeyDefinition.Default.DisplayString());
+
+			resetHotkeyButton = hotkeyDialogRoot.Get<ButtonWidget>("RESET_HOTKEY_BUTTON");
+			resetHotkeyButton.OnClick = ResetHotkey;
+
+			clearHotkeyButton = hotkeyDialogRoot.Get<ButtonWidget>("CLEAR_HOTKEY_BUTTON");
+			clearHotkeyButton.OnClick = ClearHotkey;
+
+			hotkeyEntryWidget = hotkeyDialogRoot.Get<HotkeyEntryWidget>("HOTKEY_ENTRY");
+			hotkeyEntryWidget.IsValid = () => isHotkeyValid;
+			hotkeyEntryWidget.OnLoseFocus = ValidateHotkey;
+		}
+
+		void ValidateHotkey()
+		{
+			duplicateHotkeyDefinition = modData.Hotkeys.GetFirstDuplicate(selectedHotkeyDefinition.Name, hotkeyEntryWidget.Key, selectedHotkeyDefinition);
+			isHotkeyValid = duplicateHotkeyDefinition == null;
+
+			duplicateHotkeyNotice.Visible = !isHotkeyValid;
+			clearHotkeyButton.Disabled = !hotkeyEntryWidget.Key.IsValid();
+
+			if (hotkeyEntryWidget.Key == selectedHotkeyDefinition.Default || (!hotkeyEntryWidget.Key.IsValid() && !selectedHotkeyDefinition.Default.IsValid()))
+			{
+				defaultHotkeyNotice.Visible = !duplicateHotkeyNotice.Visible;
```

Comment:
We generally prefer to make widgets observe other state rather than toggle visibility/disabled/etc manually. This really helps avoid edge cases and weird UI bugs.

In these cases it could be good to define a `enum HotkeyStatus { Default, Custom, Duplicate, Invalid }` (not sure exactly what cases we need) and an `HotkeyStatus activeHotkeyStatus;`. We can then bind e.g. `resetHotkeyButton.IsDisabled = () => activeHotkeyStatus != HotkeyStatus.Default` and so on.

This has a nice side-effect of keeping the button / label widgets local, so they won't need to be defined at the class scope.

Reply:
I found that a single enum did not cut it because a hotkey can be `Default` (or not) and `Duplicate` at the same time. I used two booleans instead - `isHotkeyValid` and `isHotkeyDefault`.